engineapi/proxy.go

```diff
@@ -1,168 +1,429 @@
+func NewEngineClientProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore, resyncPeriod time.Duration, stopCh chan struct{}) *EngineClientProxyHandler {
 	return &EngineClientProxyHandler{
-		Clients: make(map[string]EngineClientProxy),
-		logger:  logger,
-		ds:      ds,
-		lock:    &sync.RWMutex{},
+		proxyPools:   make(map[string]engineClientProxyPool),
+		logger:       logger,
+		ds:           ds,
+		stopCh:       stopCh,
+		lock:         &sync.RWMutex{},
+		resyncPeriod: resyncPeriod,
 	}
 }
 
-func getLoggerForProxy(logger logrus.FieldLogger, image string) *logrus.Entry {
+func getLoggerForProxyHandler(logger logrus.FieldLogger, im *longhorn.InstanceManager) *logrus.Entry {
 	return logger.WithFields(
 		logrus.Fields{
-			"instanceManagerImage": image,
+			"instanceManager": im.Name,
+			"image":           im.Spec.Image,
 		},
 	)
 }
 
+func (h *EngineClientProxyHandler) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	h.logger.Info("Starting Longhorn engine client proxy handler")
+	defer h.logger.Info("Shutting down Longhorn engine client proxy handler")
+
+	go wait.Until(h.resync, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (h *EngineClientProxyHandler) resync() {
+	for {
+		handledPools := h.getHandledEngineClientProxyPools()
+		for _, pool := range handledPools {
+			im, err := h.ds.GetInstanceManagerRO(pool.instanceManagerName)
+			if err != nil {
+				if !datastore.ErrorIsNotFound(err) {
+					h.logger.WithError(err).Warnf("Failed to get %v instance manager for engine client proxy", pool.instanceManagerName)
+					continue
+				}
+
+				h.logger.Warnf("%v instance manager no longer exist for engine client proxy", pool.instanceManagerName)
+				for _, client := range pool.clients {
+					engineClientProxy := client.getEngineClientProxy()
+					engineClientProxy.setNextGarbageCollect()
+				}
+				continue
+			}
+
+			for clientUUID, client := range pool.clients {
+				engineClientProxy := client.getEngineClientProxy()
+				if !engineClientProxy.shouldResync() {
+					continue
+				}
+
+				log := engineClientProxy.getLogger()
+
+				// Try to stop and delete clients thats marked for GC. We will
+				// force delete after some retry if the client cannot be stopped
+				// or the calls does not return by the caller.
+				h.garbageCollectEngineClientProxy(pool.instanceManagerName, clientUUID)
+
+				// Replace client in bad connection state with new client. Mark
+				// the bad client for GC so it will not be called for any new
+				// requests.
+				if engineClientProxy.shouldGarbageCollect() {
+					engineClientProxy.setNextGarbageCollect()
+
+					if err := h.NewEngineClientProxy(im); err != nil {
+						log.WithError(err).Debug("Failed to replace with new engine client proxy")
+					}
+
+					engineClientProxy.setNextResync(h.resyncPeriod)
+					break
+				}
+
+				// Request to create a new client
+				if err := engineClientProxy.start(im, h.logger, h.ds, h.stopCh); err != nil {
```

Comment:
This comment and the below `engineClientProxy.start` are confusing...
`engineClientProxy` blindly executing `start` really confuse me... it's better to rename this method. I was thinking this will be executed during init or connection retries...